src/box/txn.c

```diff
@@ -1069,65 +1069,41 @@ txn_add_limbo_entry(struct txn *txn, const struct journal_entry *req)
 	return 0;
 }
 
-int
-txn_commit_try_async(struct txn *txn)
-{
-	struct journal_entry *req;
-
-	ERROR_INJECT(ERRINJ_TXN_COMMIT_ASYNC, {
-		diag_set(ClientError, ER_INJECTION,
-			 "txn commit async injection");
-		goto rollback;
-	});
-
-	if (txn_prepare(txn) != 0)
-		goto rollback;
-
-	if (txn_commit_nop(txn))
-		return 0;
-
-	req = txn_journal_entry_new(txn);
-	if (req == NULL)
-		goto rollback;
-
-	if (txn_has_flag(txn, TXN_WAIT_SYNC) &&
-	    txn_add_limbo_entry(txn, req) != 0) {
-		goto rollback;
-	}
-
-	fiber_set_txn(fiber(), NULL);
-	if (journal_write_try_async(req) != 0) {
-		fiber_set_txn(fiber(), txn);
-		diag_log();
-		goto rollback;
-	}
-
-	return 0;
-
-rollback:
-	assert(txn->fiber == NULL);
-	txn_abort(txn);
-	return -1;
-}
-
-int
-txn_commit(struct txn *txn)
+static int
+txn_commit_impl(struct txn *txn, enum txn_commit_wait_mode wait_mode)
 {
 	struct journal_entry *req;
-
 	txn->fiber = fiber();
-
 	if (txn_prepare(txn) != 0)
 		goto rollback_abort;
-
 	if (txn_commit_nop(txn)) {
 		txn_free(txn);
 		return 0;
 	}
-
 	req = txn_journal_entry_new(txn);
 	if (req == NULL)
 		goto rollback_abort;
+	if (wait_mode != TXN_COMMIT_WAIT_MODE_COMPLETE) {
+		assert(wait_mode == TXN_COMMIT_WAIT_MODE_SUBMIT);
+		ERROR_INJECT(ERRINJ_TXN_COMMIT_ASYNC, {
+			diag_set(ClientError, ER_INJECTION,
+				 "txn commit async injection");
+			goto rollback_abort;
+		});
+		/*
+		 * The main reason why it is disabled is that the sync txn's
+		 * original instance wants to write CONFIRM. But it can't be
+		 * done asynchronously in the current code design. Also it is
+		 * logically strange to commit a synchronous transaction in an
+		 * async way, but the CONFIRM problem is the technical one.
+		 */
+		if (txn_has_flag(txn, TXN_WAIT_ACK) &&
```

Comment:
Maybe we can move this check to the beginning of the `txn_commit_impl`? I see no sense to prepare the tx if we'll anyway fail. Also the errinj was at the beginning, though it seems to work ok here too

Reply:
If I put this before prepare, then it would also be before `txn_journal_entry_new()`. This one fills `journal_entry` with necessary stuff, like `rows[0]->replica_id` and various flags. So if I move this before preparation, then I also have to create the journal entry before preparation. It seems worse, because prepare can fail during regular runtime if the txn was aborted by a conflict, for example, while this error right here only happens as a programming error in the application code, i.e. it mustn't happen at all, ideally.

The motivation was that we better optimize the more common runtime error path and leave the code error path in a somewhat suboptimal state.